R/separate-wider.R

```diff
@@ -0,0 +1,598 @@
+    cli::cli_abort("{.arg names} must be an unnamed character vector.")
+  }
+  too_few <- arg_match(too_few)
+  too_many <- arg_match(too_many)
+  check_bool(col_remove)
+
+  error_call %<~% current_env()
+
+  map_unpack(
+    data, {{ cols }},
+    function(x, col) str_separate_wider_delim(x, col,
+      names = names,
+      delim = delim,
+      names_sep = names_sep,
+      too_few = too_few,
+      too_many = too_many,
+      col_remove = col_remove,
+      error_call = error_call
+    ),
+    names_sep = names_sep,
+    names_repair = names_repair
+  )
+}
+
+str_separate_wider_delim <- function(
+    x,
+    col,
+    names,
+    delim,
+    names_sep = NULL,
+    too_few = "error",
+    too_many = "error",
+    col_remove = TRUE,
+    error_call = caller_env()
+) {
+
+  if (is_bare_string(delim)) {
+    delim <- stringr::fixed(delim)
+  }
+  if (too_many == "merge") {
+    if (is.null(names)) {
+      cli::cli_abort(
+        'Must provide {.arg names} when {.code too_many = "merge"}.',
+        call = error_call
+      )
+    }
+    n <- length(names)
+  } else {
+    n <- Inf
+  }
+
+  pieces <- stringr::str_split(x, delim, n = n)
+  n_pieces <- ifelse(is.na(x), NA, lengths(pieces))
+
+  names <- names %||% as.character(seq_len(max(lengths(pieces))))
+  p <- length(names)
+
+  check_df_alignment(col, p, "pieces", n_pieces,
+    too_few = too_few,
+    too_many = too_many,
+    advice_short = c(
+      i = 'Use `too_few = "debug"` to diagnose the problem.',
+      i = 'Use `too_few = "align_start"/"align_end"` to silence this message.'
+    ),
+    advice_long = c(
+      i = 'Use `too_many = "debug"` to diagnose the problem.',
+      i = 'Use `too_many = "drop"/"merge"` to silence this message.'
+    ),
+    call = error_call
+  )
+
+  out <- df_align(
+    x = pieces,
+    names = names,
+    align_direction = if (too_few == "align_end") "end" else "start"
+  )
+
+  if (!col_remove || too_few == "debug" || too_many == "debug") {
+    out[[col]] <- x
+  }
+
+  if (too_few == "debug" || too_many == "debug") {
+    separate_warn_debug(col, names_sep, c("ok", "pieces", "remainder"))
+    sep_loc <- stringr::str_locate_all(x, delim)
+    sep_last <- lapply(sep_loc, function(x) if (nrow(x) < p) NA else x[p, "start"])
+    remainder <- stringr::str_sub(x, sep_last)
+    remainder[is.na(remainder) & !is.na(x)] <- ""
+
+    problem <- !is.na(x) & (
+        (too_few == "debug" & n_pieces < p) |
+        (too_many == "debug" & n_pieces > p)
+      )
+
+    out[[debug_name(col, names_sep, "ok")]] <- !problem
+    out[[debug_name(col, names_sep, "pieces")]] <- n_pieces
+    out[[debug_name(col, names_sep, "remainder")]] <- remainder
+  }
+
+  out
+}
+
+#' @rdname separate_wider_delim
+#' @param widths A named numeric vector where the names become column names,
+#'   and the values specify the column width. Unnamed components will match,
+#'   but not be included in the output.
+#' @export
+separate_wider_position <- function(
```

Comment:
Something feels wrong with the name `_position`. When I do this:

```r
> separate_wider_position(df, x, c(z = 2, w = 3))
# A tibble: 3 × 3
     id z     w    
  <int> <chr> <chr>
1     1 m-    123  
2     2 f-    455  
3     3 f-    123  
```

I am supplying _sizes_ (i.e. _widths_) not positions. This is most apparent with `w = 3`, where `3` is a width, not position 3 in the string - it is really positions 3-5 in the string because the positions is based on the cumulative sizes of the widths

`separate_wider_width()`? `separate_wider_size()`? Tough call on an alternate name

Reply:
Yeah, we talked about this at at the group meeting. `position` isn't perfect but it's not as bad as `separate_wider_width`

Reply:
sounds good, i figured as much